File: sql/udf_framework/python/udf_helpers.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
ExaPG UDF Helpers für Python
Diese Bibliothek bietet Exasol-kompatible Funktionen für Python-UDFs in PostgreSQL
"""

import json
import datetime
import pandas as pd
import numpy as np
import plpy  # PostgreSQL-Python-Schnittstelle
import re

# Exasol-Kompatibilitätsschicht
class Exasol:
    """
    Exasol-Kompatibilitätsschicht für PostgreSQL Python UDFs
    """
    
    class Connection:
        """Simuliert eine Exasol-Datenbankverbindung in PostgreSQL"""
        
        def __init__(self, connection_name):
            self.connection_name = connection_name
        
        def execute(self, query, *args):
            """Führt eine SQL-Query aus mit Exasol-ähnlicher Parameterisierung"""
            # Konvertiere Exasol-Style (?)-Parameter zu PostgreSQL $-Parameter
            param_counter = 1
            converted_query = ""
            for char in query:
                if char == '?':
                    converted_query += f"${param_counter}"
                    param_counter += 1
                else:
                    converted_query += char
            
            # Führe die Query aus
            result = plpy.execute(converted_query, args)
            return [[col for col in row.values()] for row in result]
        
        def commit(self):
            """Commit-Operation (in PostgreSQL UDFs implizit)"""
            # In PostgreSQL PL/Python wird automatisch committed
            pass
        
        def rollback(self):
            """Rollback-Operation (limitiert in UDFs)"""
            # In PostgreSQL PL/Python sind Rollbacks begrenzt möglich
            plpy.execute("ROLLBACK")
        
        def close(self):
            """Schließt die Verbindung (in PostgreSQL UDFs implizit)"""
            # In PostgreSQL PL/Python sind Verbindungen Teil der Session
            pass
    
    def __init__(self):
        """Initialisiert die Exasol-Kompatibilitätsschicht"""
        pass
    
    def get_connection(self, connection_name="default"):
        """Gibt ein Connection-Objekt zurück"""
        return self.Connection(connection_name)
    
    # Logging-Funktionen
    def error_msg(self, message):
        """Log eine Fehlermeldung"""
        plpy.error(message)
    
    def info_msg(self, message):
        """Log eine Informationsmeldung"""
        plpy.notice(message)
    
    def debug_msg(self, message):
        """Log eine Debug-Meldung"""
        plpy.notice(f"[DEBUG] {message}")
    
    # Datentyp-Konvertierung
    def to_timestamp(self, date_str):
        """Konvertiert einen String zu einem Timestamp"""
        if date_str is None:
            return None
        try:
            # Wenn es bereits ein Datetime-Objekt ist
            if isinstance(date_str, (datetime.datetime, datetime.date)):
                return date_str
            # Versuche verschiedene Formate
            for fmt in ('%Y-%m-%d %H:%M:%S', '%Y-%m-%d', '%d.%m.%Y %H:%M:%S', '%d.%m.%Y'):
                try:
                    return datetime.datetime.strptime(date_str, fmt)
                except ValueError:
                    continue
            raise ValueError(f"Ungültiges Datumsformat: {date_str}")
        except Exception as e:
            self.error_msg(f"Fehler bei der Konvertierung zu Timestamp: {e}")
            return None
    
    def to_date(self, date_str):
        """Konvertiert einen String zu einem Date"""
        if date_str is None:
            return None
        try:
            # Wenn es bereits ein Date-Objekt ist
            if isinstance(date_str, datetime.date):
                return date_str
            # Versuche verschiedene Formate
            for fmt in ('%Y-%m-%d', '%d.%m.%Y'):
                try:
                    return datetime.datetime.strptime(date_str, fmt).date()
                except ValueError:
                    continue
            raise ValueError(f"Ungültiges Datumsformat: {date_str}")
        except Exception as e:
            self.error_msg(f"Fehler bei der Konvertierung zu Date: {e}")
            return None
    
    def to_number(self, value):
        """Konvertiert einen Wert zu einer Zahl"""
        if value is None:
            return None
        try:
            return float(value)
        except Exception as e:
            self.error_msg(f"Fehler bei der Konvertierung zu Number: {e}")
            return None
    
    def to_char(self, value):
        """Konvertiert einen Wert zu einem String"""
        if value is None:
            return None
        return str(value)
    
    # JSON-Verarbeitung
    class Json:
        @staticmethod
        def encode(value):
            """Konvertiert einen Python-Wert zu einem JSON-String"""
            try:
                return json.dumps(value)
            except Exception as e:
                plpy.error(f"JSON-Enkodierungsfehler: {e}")
                return None
        
        @staticmethod
        def decode(json_str):
            """Konvertiert einen JSON-String zu einem Python-Wert"""
            try:
                return json.loads(json_str)
            except Exception as e:
                plpy.error(f"JSON-Dekodierungsfehler: {e}")
                return None
    
    # Mathematische Funktionen
    class Math:
        @staticmethod
        def round(value, digits=0):
            """Rundet einen Wert auf die angegebene Anzahl von Dezimalstellen"""
            if value is None:
                return None
            return round(value, digits)
        
        @staticmethod
        def floor(value):
            """Rundet einen Wert ab"""
            if value is None:
                return None
            return np.floor(value)
        
        @staticmethod
        def ceil(value):
            """Rundet einen Wert auf"""
            if value is None:
                return None
            return np.ceil(value)
        
        @staticmethod
        def abs(value):
            """Gibt den Absolutwert zurück"""
            if value is None:
                return None
            return abs(value)
        
        @staticmethod
        def sqrt(value):
            """Gibt die Quadratwurzel zurück"""
            if value is None or value < 0:
                return None
            return np.sqrt(value)
        
        @staticmethod
        def mean(values):
            """Berechnet den Mittelwert einer Liste von Werten"""
            if not values:
                return None
            return np.mean(values)
        
        @staticmethod
        def median(values):
            """Berechnet den Median einer Liste von Werten"""
            if not values:
                return None
            return np.median(values)
        
        @staticmethod
        def stdev(values):
            """Berechnet die Standardabweichung einer Liste von Werten"""
            if not values or len(values) < 2:
                return None
            return np.std(values, ddof=1)  # ddof=1 für Stichproben-Standardabweichung
    
    # String-Funktionen
    class Text:
        @staticmethod
        def lower(text):
            """Konvertiert einen String zu Kleinbuchstaben"""
            if text is None:
                return None
            return text.lower()
        
        @staticmethod
        def upper(text):
            """Konvertiert einen String zu Großbuchstaben"""
            if text is None:
                return None
            return text.upper()
        
        @staticmethod
        def substring(text, start, length=None):
            """Extrahiert einen Teilstring"""
            if text is None:
                return None
            # Exasol verwendet 1-basierte Indizierung, Python verwendet 0-basierte
            start_idx = start - 1
            if length is None:
                return text[start_idx:]
            else:
                return text[start_idx:start_idx + length]
        
        @staticmethod
        def replace(text, search, replace):
            """Ersetzt alle Vorkommen eines Substrings"""
            if text is None:
                return None
            return text.replace(search, replace)
        
        @staticmethod
        def trim(text):
            """Entfernt Leerzeichen am Anfang und Ende"""
            if text is None:
                return None
            return text.strip()
        
        @staticmethod
        def left(text, n):
            """Gibt die ersten n Zeichen zurück"""
            if text is None:
                return None
            return text[:n]
        
        @staticmethod
        def right(text, n):
            """Gibt die letzten n Zeichen zurück"""
            if text is None:
                return None
            return text[-n:]
        
        @staticmethod
        def length(text):
            """Gibt die Länge eines Strings zurück"""
            if text is None:
                return None
            return len(text)
        
        @staticmethod
        def regex_replace(text, pattern, replacement, flags=0):
            """Ersetzt Text basierend auf einem regulären Ausdruck"""
            if text is None:
                return None
            return re.sub(pattern, replacement, text, flags=flags)
    
    # Datum/Zeit-Funktionen
    class Date:
        @staticmethod
        def now():
            """Gibt den aktuellen Zeitstempel zurück"""
            return datetime.datetime.now()
        
        @staticmethod
        def today():
            """Gibt das aktuelle Datum zurück"""
            return datetime.date.today()
        
        @staticmethod
        def add_days(date, days):
            """Fügt Tage zu einem Datum hinzu"""
            if date is None:
                return None
            return date + datetime.timedelta(days=days)
        
        @staticmethod
        def add_months(date, months):
            """Fügt Monate zu einem Datum hinzu"""
            if date is None:
                return None
            
            # Berechne das neue Jahr und den neuen Monat
            year = date.year + (date.month + months - 1) // 12
            month = (date.month + months - 1) % 12 + 1
            
            # Behalte den gleichen Tag, wenn möglich, ansonsten letzter Tag des Monats
            day = min(date.day, [31, 29 if year % 4 == 0 and (year % 100 != 0 or year % 400 == 0) else 28, 
                               31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month-1])
            
            if isinstance(date, datetime.datetime):
                return datetime.datetime(year, month, day, date.hour, date.minute, date.second, date.microsecond)
            else:
                return datetime.date(year, month, day)
        
        @staticmethod
        def add_years(date, years):
            """Fügt Jahre zu einem Datum hinzu"""
            if date is None:
                return None
            
            # Beachte Schaltjahre
            new_year = date.year + years
            new_month = date.month
            
            # Wenn es der 29. Februar ist und das neue Jahr kein Schaltjahr ist
            if date.month == 2 and date.day == 29 and not (new_year % 4 == 0 and (new_year % 100 != 0 or new_year % 400 == 0)):
                new_day = 28
            else:
                new_day = date.day
            
            if isinstance(date, datetime.datetime):
                return datetime.datetime(new_year, new_month, new_day, date.hour, date.minute, date.second, date.microsecond)
            else:
                return datetime.date(new_year, new_month, new_day)
        
        @staticmethod
        def diff_days(date1, date2):
            """Berechnet die Differenz in Tagen zwischen zwei Daten"""
            if date1 is None or date2 is None:
                return None
            delta = date1 - date2
            return delta.days
        
        @staticmethod
        def diff_months(date1, date2):
            """Berechnet die Differenz in Monaten zwischen zwei Daten"""
            if date1 is None or date2 is None:
                return None
            return (date1.year - date2.year) * 12 + (date1.month - date2.month)
        
        @staticmethod
        def diff_years(date1, date2):
            """Berechnet die Differenz in Jahren zwischen zwei Daten"""
            if date1 is None or date2 is None:
                return None
            return date1.year - date2.year
    
    # Metadaten-Funktionen
    class Meta:
        @staticmethod
        def schema_name():
            """Gibt den aktuellen Schema-Namen zurück"""
            try:
                result = plpy.execute("SELECT current_schema() AS schema")[0]
                return result["schema"]
            except Exception as e:
                plpy.error(f"Fehler beim Abrufen des Schema-Namens: {e}")
                return None
        
        @staticmethod
        def current_user():
            """Gibt den aktuellen Benutzernamen zurück"""
            try:
                result = plpy.execute("SELECT current_user AS user")[0]
                return result["user"]
            except Exception as e:
                plpy.error(f"Fehler beim Abrufen des Benutzernamens: {e}")
                return None
        
        @staticmethod
        def current_session():
            """Gibt die aktuelle Session-ID zurück"""
            try:
                result = plpy.execute("SELECT pg_backend_pid() AS pid")[0]
                return result["pid"]
            except Exception as e:
                plpy.error(f"Fehler beim Abrufen der Session-ID: {e}")
                return None
    
    # DataFrame-Unterstützung für Pandas
    class DataFrame:
        @staticmethod
        def from_query(query, conn, *params):
            """Erstellt einen Pandas-DataFrame aus einer SQL-Abfrage"""
            try:
                # Führe die Abfrage aus
                result = conn.execute(query, *params)
                
                # Wenn keine Daten zurückgegeben wurden
                if not result:
                    return pd.DataFrame()
                
                # Erstelle einen DataFrame
                return pd.DataFrame(result)
            except Exception as e:
                plpy.error(f"Fehler beim Erstellen des DataFrame: {e}")
                return pd.DataFrame()
        
        @staticmethod
        def to_query(df, table_name, conn, if_exists='replace'):
            """Schreibt einen Pandas-DataFrame in eine Tabelle"""
            try:
                # Konvertiere DataFrame zu einer Liste von Tupeln
                records = df.to_records(index=False)
                
                # Erstelle die Tabelle, falls sie nicht existiert
                if if_exists == 'replace':
                    # Erzeuge SQL für Tabellen-Drop
                    conn.execute(f"DROP TABLE IF EXISTS {table_name}")
                
                # Erzeuge Tabelle mit passenden Spaltentypen
                columns = []
                for col_name, dtype in zip(df.columns, df.dtypes):
                    pg_type = "TEXT"  # Standard-Typ
                    
                    # Mapping von Pandas-Typen zu PostgreSQL-Typen
                    if np.issubdtype(dtype, np.integer):
                        pg_type = "INTEGER"
                    elif np.issubdtype(dtype, np.floating):
                        pg_type = "DOUBLE PRECISION"
                    elif np.issubdtype(dtype, np.datetime64):
                        pg_type = "TIMESTAMP"
                    elif np.issubdtype(dtype, np.bool_):
                        pg_type = "BOOLEAN"
                    
                    columns.append(f"{col_name} {pg_type}")
                
                # Erstelle die Tabelle
                create_query = f"CREATE TABLE {table_name} ({', '.join(columns)})"
                conn.execute(create_query)
                
                # Füge Daten ein
                for record in records:
                    placeholders = ", ".join(["?"] * len(record))
                    insert_query = f"INSERT INTO {table_name} VALUES ({placeholders})"
                    conn.execute(insert_query, *record)
                
                return True
            except Exception as e:
                plpy.error(f"Fehler beim Schreiben des DataFrame: {e}")
                return False

# Erstelle eine globale Instanz der Exasol-Klasse
exa = Exasol()

# Setze Komponenten als direkte Attribute für einfacheren Zugriff
exa.json = exa.Json
exa.math = exa.Math
exa.text = exa.Text
exa.date = exa.Date
exa.meta = exa.Meta
exa.df = exa.DataFrame

# Exportiere das Exasol-Objekt
__all__ = ['exa'] 

File: sql/udf_framework/lua/udf_helpers.lua
-- ExaPG UDF Helpers für Lua
-- Diese Bibliothek bietet Exasol-kompatible Funktionen für Lua-UDFs in PostgreSQL

-- Globales Objekt für PostgreSQL-Funktionalität
pg = pg or {}

-- Exasol-Kompatibilitätsschicht
exasol = {}

-- Logging-Funktionen
exasol.error_msg = function(message)
    pg.error(message)
end

exasol.info_msg = function(message)
    pg.notice(message)
end

exasol.debug_msg = function(message)
    -- In PostgreSQL ist debug-level-logging nicht direkt verfügbar,
    -- daher verwenden wir notice mit Präfix
    pg.notice("[DEBUG] " .. message)
end

-- Datenbankfunktionen
exasol.get_connection = function(connection_name)
    -- Simuliert Exasols Verbindungsverwaltung
    return {
        name = connection_name,
        -- Ausführen einer SQL-Query mit parametrisierten Statements
        execute = function(self, query, ...)
            -- Konvertiere Exasol-Style-SQL zu PostgreSQL-kompatiblem SQL
            local converted_query = query:gsub(":%d+", function(match)
                local param_num = tonumber(match:sub(2))
                return "$" .. param_num
            end)
            
            -- Führe die Query aus
            return pg.execute(converted_query, {...})
        end,
        -- Schließt die Verbindung
        close = function(self)
            -- In PostgreSQL benötigen wir kein explizites Schließen
            return true
        end
    }
end

-- Datentyp-Konvertierung
exasol.to_timestamp = function(date_str)
    return pg.execute("SELECT TO_TIMESTAMP($1, 'YYYY-MM-DD HH24:MI:SS')", {date_str})[1][1]
end

exasol.to_date = function(date_str)
    return pg.execute("SELECT TO_DATE($1, 'YYYY-MM-DD')", {date_str})[1][1]
end

exasol.to_number = function(value)
    if value == nil then return nil end
    return tonumber(value)
end

exasol.to_char = function(value)
    if value == nil then return nil end
    return tostring(value)
end

-- JSON-Verarbeitung
exasol.json = {
    encode = function(value)
        -- Nutze die json-Bibliothek von Lua
        local json = require("json")
        return json.encode(value)
    end,
    
    decode = function(json_str)
        local json = require("json")
        local status, result = pcall(json.decode, json_str)
        if status then
            return result
        else
            exasol.error_msg("JSON Decode-Fehler: " .. result)
            return nil
        end
    end
}

-- Mathematische Funktionen
exasol.math = {
    round = function(value, digits)
        digits = digits or 0
        local factor = 10 ^ digits
        return math.floor(value * factor + 0.5) / factor
    end,
    
    floor = math.floor,
    ceil = math.ceil,
    abs = math.abs,
    sqrt = math.sqrt,
    
    -- Statistische Funktionen
    mean = function(values)
        if #values == 0 then return nil end
        local sum = 0
        for _, v in ipairs(values) do
            sum = sum + v
        end
        return sum / #values
    end,
    
    median = function(values)
        if #values == 0 then return nil end
        
        -- Kopiere die Tabelle, um die Originalwerte nicht zu verändern
        local sorted = {}
        for i, v in ipairs(values) do
            sorted[i] = v
        end
        table.sort(sorted)
        
        local n = #sorted
        if n % 2 == 0 then
            return (sorted[n/2] + sorted[n/2 + 1]) / 2
        else
            return sorted[math.ceil(n/2)]
        end
    end,
    
    stdev = function(values)
        if #values < 2 then return nil end
        
        -- Berechne Mittelwert
        local mean = exasol.math.mean(values)
        
        -- Berechne Summe der quadratischen Abweichungen
        local sum_sq_diff = 0
        for _, v in ipairs(values) do
            sum_sq_diff = sum_sq_diff + (v - mean)^2
        end
        
        -- Berechne Standardabweichung
        return math.sqrt(sum_sq_diff / (#values - 1))
    end
}

-- String-Funktionen
exasol.text = {
    lower = string.lower,
    upper = string.upper,
    
    substring = function(str, start_pos, length)
        if str == nil then return nil end
        
        -- Exasol-Indizierung beginnt bei 1, wie in Lua
        if length then
            return string.sub(str, start_pos, start_pos + length - 1)
        else
            return string.sub(str, start_pos)
        end
    end,
    
    replace = function(str, search, replace)
        if str == nil then return nil end
        return string.gsub(str, search, replace)
    end,
    
    trim = function(str)
        if str == nil then return nil end
        return string.match(str, "^%s*(.-)%s*$")
    end,
    
    left = function(str, n)
        if str == nil then return nil end
        return string.sub(str, 1, n)
    end,
    
    right = function(str, n)
        if str == nil then return nil end
        return string.sub(str, -n)
    end,
    
    length = function(str)
        if str == nil then return nil end
        return string.len(str)
    end
}

-- Datum/Zeit-Funktionen
exasol.date = {
    now = function()
        return pg.execute("SELECT NOW()")[1][1]
    end,
    
    add_days = function(date, days)
        return pg.execute("SELECT $1 + INTERVAL '$2 days'", {date, days})[1][1]
    end,
    
    add_months = function(date, months)
        return pg.execute("SELECT $1 + INTERVAL '$2 months'", {date, months})[1][1]
    end,
    
    add_years = function(date, years)
        return pg.execute("SELECT $1 + INTERVAL '$2 years'", {date, years})[1][1]
    end,
    
    diff_days = function(date1, date2)
        return pg.execute("SELECT DATE_PART('day', $1 - $2)", {date1, date2})[1][1]
    end,
    
    diff_months = function(date1, date2)
        return pg.execute("SELECT (DATE_PART('year', $1) - DATE_PART('year', $2)) * 12 + (DATE_PART('month', $1) - DATE_PART('month', $2))", {date1, date2})[1][1]
    end,
    
    diff_years = function(date1, date2)
        return pg.execute("SELECT DATE_PART('year', $1) - DATE_PART('year', $2)", {date1, date2})[1][1]
    end,
    
    extract = function(part, date)
        return pg.execute("SELECT DATE_PART($1, $2)", {part, date})[1][1]
    end
}

-- Metadaten-Funktionen
exasol.meta = {
    -- Simuliert Exasols Funktionen zur Abfrage von Metadaten
    schema_name = function()
        return pg.execute("SELECT CURRENT_SCHEMA()")[1][1]
    end,
    
    script_name = function()
        -- In PostgreSQL gibt es kein direktes Äquivalent zu script_name
        -- Verwende den Funktionsnamen als Ersatz
        return pg.execute("SELECT pg_proc.proname FROM pg_proc WHERE pg_proc.oid = pg_proc.oid")[1][1]
    end,
    
    current_user = function()
        return pg.execute("SELECT CURRENT_USER")[1][1]
    end,
    
    current_session = function()
        -- Simulierte Session-ID in PostgreSQL
        return pg.execute("SELECT pg_backend_pid()")[1][1]
    end
}

-- Tabellenfunktionen
exasol.table = {
    new = function()
        return {}
    end,
    
    insert = function(tbl, row)
        table.insert(tbl, row)
        return true
    end,
    
    column = function(tbl, col_idx)
        local result = {}
        for i, row in ipairs(tbl) do
            result[i] = row[col_idx]
        end
        return result
    end,
    
    row_count = function(tbl)
        return #tbl
    end
}

-- Exportiere Exasol-Namespace für Kompatibilität
return exasol 